### PR TITLE
refactor(metadata): use async/await

### PR DESCRIPTION
**Pull request by ofrobots**
Get rid of unnecessary class and use async/await.

### CODE REVIEW

File: src/log.ts [hunk 1]
```diff
@@ -640,10 +640,15 @@ class Log {
         options = {};
       }
       if (!options.resource) {
-        this.metadata_.getDefaultResource((err, resource) => {
-          // Ignore errors (the API will speak up if it has an issue).
-          writeWithResource(resource);
-        });
+        getDefaultResource(this.logging.auth)
```
[the remark this answers is not in the document]
Comment (ofrobots) [reply src/log.ts]: which constructor?

File: src/log.ts [hunk 2]
```diff
@@ -20,10 +20,12 @@ import * as arrify from 'arrify';
 import {promisifyAll} from '@google-cloud/promisify';
 import * as extend from 'extend';
 import * as is from 'is';
+
```
[the remark this answers is not in the document]
Comment (ofrobots) [reply src/log.ts]: This extra new line makes sense as it separates the `node_module` imports from local imports. Unwritten rule that I follow. I would be happy to put it back though.

File: src/log.ts [hunk 3]
```diff
@@ -50,14 +52,12 @@ const {Metadata} = require('./metadata');
 class Log {
   formattedName_: string;
   removeCircular_: boolean;
-  metadata_;
```
[the remark this answers is not in the document]
Comment (ofrobots) [reply src/log.ts]: The `Log` class is exported and instances are part of the public API as well. There should be no uses of `metadata_` in the ecosystem. `logging-{winston,bunyan}` certainly don't use it.

File: test/metadata.ts [hunk 4]
```diff
@@ -145,158 +135,134 @@ describe('metadata', () => {
...
-    it('should return error on failure to acquire metadata', done => {
+    it('should throw error on failure to acquire metadata', async () => {
+      let caught = false;
       const FAKE_ERROR = new Error();
       instanceOverride = {
         errorArg: FAKE_ERROR,
       };
 
-      Metadata.getGKEDescriptor(err => {
+      try {
```
[the remark this answers is not in the document]
Comment (ofrobots) [reply test/metadata.ts]: AFAICT, there are no real dependents other than the author's own modules.

I would be happy to pick something that was standard solution.

File: package.json [hunk 5]
```diff
@@ -83,6 +83,7 @@
     "is": "^3.2.1",
     "lodash.merge": "^4.6.1",
     "on-finished": "^2.3.0",
+    "pify": "^4.0.1",
```
[the remark this answers is not in the document]
Comment (ofrobots) [reply package.json]: fixed

File: test/metadata.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (ofrobots) [reply test/metadata.ts]: Looked at the contents. It seemed fine.

[the remark this answers is not in the document]
Comment (JustinBeckwith) [reply test/metadata.ts]: Yeah, I'm using it in google-auth-library FWIW and it did what it says it do 🤷‍♂️   `assert.rejects` is only available in nodejs 10, so I think we got a while to wait.

**Pull request merged**